lib/BloqadeQMC/src/updates/cluster.jl

```diff
@@ -0,0 +1,333 @@
+    leg_sites = qmc_state.leg_sites
+
+    if qmc_state isa BinaryGroundState
+        First = qmc_state.first
+
+        # The first N elements of the linked list are the spins of the LHS basis state
+        @inbounds for i in 1:Ns
+            LegType[i] = spin_left[i]
+            Associates[i] = 0
+            First[i] = i
+            if H isa AbstractLTFIM
+                op_indices[i] = 0
+                leg_sites[i] = 0
+            end
+        end
+        idx = Ns
+    else
+        First = fill!(qmc_state.first, 0)  #initialize the First list
+        Last = fill!(qmc_state.last, 0)   #initialize the Last list
+        idx = 0
+    end
+
+    spin_prop = copyto!(qmc_state.propagated_config, spin_left)  # the propagated spin state
+
+    # Now, add the 2M operators to the linked list. Each has either 2 or 4 legs
+    @inbounds for (n, op) in enumerate(qmc_state.operator_list)
+        if issiteoperator(H, op)
+            site = getsite(H, op)
+            # lower or left leg
+            idx += 1
+            F = First[site]
+            LinkList[idx] = F
+            if qmc_state isa BinaryGroundState || !iszero(F)
+                LinkList[F] = idx  # completes backwards link
+            else
+                Last[site] = idx
+            end
+
+            LegType[idx] = spin_prop[site]
+            Associates[idx] = 0
+            if H isa AbstractLTFIM
+                op_indices[idx] = n
+                leg_sites[idx] = 1
+            end
+
+            if !isdiagonal(H, op)  # off-diagonal site operator
+                spin_prop[site] ⊻= 1  # spinflip
+            end
+
+            # upper or right leg
+            idx += 1
+            First[site] = idx
+            LegType[idx] = spin_prop[site]
+            Associates[idx] = 0
+            if H isa AbstractLTFIM
+                op_indices[idx] = n
+                leg_sites[idx] = 1
+            end
+        elseif qmc_state isa BinaryGroundState || isbondoperator(H, op)  # diagonal bond operator
+            site1, site2 = bond = getbondsites(H, op)
+            spins = spin_prop[site1], spin_prop[site2]
+            num_sites = 2  # length(bond)
+            num_legs = 4   # 2*num_sites
+
+            # vertex leg indices
+            # thinking of imaginary time as increasing as we move upward,
+            # these indices refer to the
+            # lower left, lower right, upper left, upper right legs respectively
+            # v1, v2, v3, v4 = idx + 1, idx + 2, idx + 3, idx + 4
+
+            @simd for i in 1:num_sites
+                v = idx + i
+                st = bond[i]
+                F = First[st]
+                LinkList[v] = F
+                if qmc_state isa BinaryGroundState || !iszero(F)
+                    LinkList[F] = v  # completes backwards link
+                else
+                    Last[st] = v
+                end
+                First[st] = v + num_sites
+            end
+
+            @simd for i in 1:num_legs
```

Comment:
Can we check if the simd is actually useful here? By measuring the time - simd optimization is not very reliable so it'd be great to check that while developing